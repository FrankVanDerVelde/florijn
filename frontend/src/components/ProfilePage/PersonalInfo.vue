

<template>

<div class="text-[20px] font-bold mb-[10px]">Profiel</div>
                <div class="name-and-picture-container flex items-center">
                    <img class="rounded-full w-[60px] h-[60px] mr-4"
                        src="https://media.istockphoto.com/vectors/thumbnail-image-vector-graphic-vector-id1147544807?k=20&m=1147544807&s=612x612&w=0&h=pBhz1dkwsCMq37Udtp9sfxbjaMl27JUapoyYpQm0anc=">
                    <div>
                        <div class="text-[20px] font-bold">Naam placeholder</div>
                        <div class="text-[14px] text-primary-400 font-bold">Title placeholder</div>
                    </div>
                </div>

  
  <div class="font-bold text-[20px]">Persoonlijke Informatie</div>

                    <div class="input-container duo">
                        <FormInput name="voornaam"></FormInput>
                        <FormInput name="achternaam"></FormInput>
                    </div>

                    <span class="adress">Woongegevens</span>
                    <div class="input-container">
                        <FormInput name="Email" type="email"></FormInput>
                    </div>

                    <div class="input-container duo">
                        <FormInput name="woonplaats"></FormInput>
                        <FormInput name="postcode"></FormInput>
                    </div>

                    <div class="input-container duo">
                        <FormInput name="huisnummer"></FormInput>
                        <FormInput name="toevoeging"></FormInput>
                    </div>

</template>
  
<style scoped>

</style>

<script>
export default {
    name: "Profile",
    data() {
        return {

        }
    }
}

</script>