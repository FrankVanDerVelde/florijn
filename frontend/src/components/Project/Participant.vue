<template>
  <div class="flex flex-row mt-4 w-full md:w-auto" v-if="!small">
    <img :src="participant.avatar" alt="Avatar" class="w-[82px] h-[82px] rounded-full mr-4">
    <div class="flex flex-col justify-between relative pr-[34px] w-full md:w-auto">
      <a class="email-btn hover:bg-neutral-50 hover:border-neutral-100 transition-all" :href="'mailto:' + participant.email">
        <font-awesome-icon icon="envelope"/>
      </a>
      <div class="flex flex-col">
        <div class="font-bold">{{ name }}</div>
        <div class="font-semibold text-neutral-500">{{ participant.role }}</div>
      </div>
      <div class="text-neutral-400">
        {{ participant.email }}
      </div>
    </div>
  </div>
  <div class="flex flex-row items-center w-full" v-else>
    <img :src="participant.avatar" alt="Avatar" class="w-[28px] h-[28px] rounded-full mr-2">
    <div class="flex flex-col justify-between">
      <div class="font-bold text-sm">{{ name }}</div>
      <div class="font-semibold text-neutral-500 text-sm">{{ participant.role }}</div>
    </div>
  </div>

</template>

<script>
export default {
  name: "Participant",

  computed: {
    name() {
      let lastNameParts = this.participant.lastName.split(" ");
      return this.participant.firstName + " " + lastNameParts[lastNameParts.length - 1].charAt(0) + ".";
    }
  },

  props: {
    participant: {
      type: Object,
      required: true
    },
    small: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style scoped>
.email-btn {
  position: absolute;
  top: 0;
  right: 0;
  color: var(--neutral-300);
  border: 1px solid var(--neutral-200);
  padding: 3px 6px;
  font-size: 14px;
  border-radius: 6px;
}
</style>