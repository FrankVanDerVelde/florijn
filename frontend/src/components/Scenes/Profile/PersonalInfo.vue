

<template>
    <form>
        <div class="text-[34px] font-bold mb-[10px]">Profiel</div>
        <div class="name-and-picture-container flex items-center mb-4">
            <div>
                <img class="rounded-full w-[80px] h-[80px] mr-4" :src="user.avatarUrl">
                <div class="profile-picture-container">
                    <label for="profile-picture-upload-input">
                        <font-awesome-icon icon="fa-file-image" class="profile-picture-upload-icon" />
                    </label>
                    <input @change="e => updateAvatar(e)" type="file" name="profile-picture-upload-input"
                        id="profile-picture-upload-input" accept=".svg,.png,.webp,jpg,.jpeg">

                </div>
            </div>
            <div>
                <div class="text-[26px] font-bold">{{ user.firstName }} {{ user.lastName }} </div>
                <div class="text-[14px] text-primary-400 font-bold">{{ user.title }}</div>
            </div>
        </div>


        <div class="grid grid-cols-2 gap-4 row-1">
            <FormInput name="voornaam" v-model="user.firstName" :validationRules="['required']"></FormInput>
            <FormInput name="achternaam" v-model="user.lastName" :validationRules="['required']"></FormInput>
            <FormInput name="Email" type="email" v-model="user.email" :validationRules="['required', 'email']">
            </FormInput>
        </div>

        <span class="text-[18px] font-bold">Woongegevens</span>


        <div class="grid grid-cols-2 gap-4 row-1">
            <FormInput name="woonplaats" v-model="address.place"></FormInput>
        </div>

        <div class="grid grid-cols-2 gap-4 row-1">
            <FormInput name="Straat" v-model="address.street"></FormInput>
        </div>

        <div class="grid grid-cols-3 gap-4">
            <FormInput name="huisnummer" type="number" v-model="address.houseNumber"></FormInput>
            <FormInput name="toevoeging" v-model="address.houseNumberAddition"></FormInput>
            <FormInput name="postcode" v-model="address.postalCode" :validationRules="['required', 'zipcode-dutch']">
            </FormInput>
        </div>

        <div class="grid grid-cols-2 gap-4 row-1">
            <SaveButton :onClick="handleProfileUpdate"></SaveButton>
        </div>
    </form>
</template>
  
<style scoped>
.input-container {
    display: flex;
    margin-bottom: 1em;
}

.input-container.duo {}

.input-container.duo>div:first-child {
    margin-right: 1em;
}

.profile-picture-container {
    position: relative;
}

.profile-picture-container label {
    position: absolute;
    color: gray;
    right: 24px;
    bottom: 0px;
    font-size: 25px;
}

.profile-picture-container input {
    display: none;
}
</style>

<script>
export default {
    name: "Profile",
    components: {
        FormInput,
        SaveButton
    },
    inject: ['userFetchService'],
    async created() {
        const newAddress = await this.userFetchService.fetchJson(`/address/${this.user.id}`);
        this.address = newAddress;
    },
    data() {
        return {
            user: JSON.parse(localStorage.getItem("user")),
            address: {},
            avatarFile: null
        }
    },
    methods: {
        async handleProfileUpdate(formValues) {
            this.user.address = this.address;
            const adjustedUser = this.user;
            ['expertises', 'password', 'role', 'skills'].forEach(key => {
                adjustedUser[key] && (delete adjustedUser[key]); 
            }) 
   
            const formData = new FormData();
            console.log(adjustedUser);
            formData.append('address', JSON.stringify(adjustedUser.address));
            delete adjustedUser.address;

            formData.append('user', JSON.stringify({
                adjustedUser
            }));

            formData.append('id', 
                adjustedUser.id
            );


            if (this.avatarFile != null) formData.append('avatarFile', this.avatarFile);

            console.log(this.avatarFile)
            this.userFetchService.fetchJsonFile(`/${this.user.id}/edit`, "PUT", formData).then((response) => {
                localStorage.setItem("user", JSON.stringify(response));
            })
        },
        async updateAvatar(event) {
            if (event.target.files.length === 0) {
                this.logoFile = null;
                delete this.user.avatarUrl;
                return;
            }

            const fileExtension = event.target.files[0].name.split('.').pop();
            if (!['svg', 'png', 'webp', 'jpg', 'jpeg'].includes(fileExtension)) {
                // this.errors.logo = "Alleen afbeeldingen met de extensies .svg, .png, .webp, .jpg en .jpeg zijn toegestaan.";
                return;
            }

            this.avatarFile = event.target.files[0];
            this.user.avatarUrl = `${await this.getBase64(event.target.files[0])}`;
        },
        async getBase64(file) {
            // convert javascript file to base64 string
            return new Promise((resolve, reject) => {
                const reader = new FileReader();
                reader.readAsDataURL(file);
                reader.onload = () => resolve(reader.result);
                reader.onerror = error => reject(error);
            });
        }
    }
}

import { VueElement } from "vue";
import FormInput from "../../Common/FormInput.vue";
import SaveButton from "../../Common/SaveButton.vue";

</script>