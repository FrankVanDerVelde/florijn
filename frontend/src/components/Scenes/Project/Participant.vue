<template>
  <div class="flex flex-row mt-4 w-full md:w-auto" v-if="!small">
    <Asset :src="participant.user.avatarUrl" alt="Avatar" class="w-[82px] h-[82px] rounded-full mr-4"/>
    <div class="flex flex-col justify-between relative pr-[34px] w-full md:w-auto">
      <div class="edit-btn hover:bg-neutral-50 hover:border-app_red-500 transition-all" v-if="edit" @click="$emit('selectedParticipant', {participant})">
        <font-awesome-icon icon="trash-can"/>
      </div>
      <a class="email-btn hover:bg-neutral-50 hover:border-neutral-100 transition-all" :href="'mailto:' + participant.user.email">
        <font-awesome-icon icon="envelope"/>
      </a>
      <div class="flex flex-col">
        <div class="font-bold">{{ name }}</div>
        <div class="font-semibold text-neutral-500">{{ participant.role }}</div>
      </div>
      <div class="text-neutral-400">
        {{ participant.user.email }}
      </div>
    </div>
  </div>
  <div class="flex flex-row items-center w-max" v-else>
    <Asset :src="participant.user.avatarUrl" alt="Avatar" class="w-[32px] h-[32px] rounded-full mr-2"/>
    <div class="flex flex-col justify-between">
      <div class="font-bold text-sm">{{ name }}</div>
      <div class="font-semibold text-neutral-500 text-sm">{{ participant.role }}</div>
    </div>

  </div>

</template>

<script>
import Asset from "../../Common/Asset.vue";

export default {
  name: "Participant",
  components: {Asset},
  emits: ["selectedParticipant"],

  created() {
    console.log(this.participant.user.avatarUrl)
  },

  computed: {
    name() {
      // user is a client
      if ('name' in this.participant.user) {
        return this.participant.user.name;
      }

      // user is a specialist/admin
      let lastNameParts = this.participant.user.lastName.split(" ");
      return this.participant.user.firstName + " " + lastNameParts[lastNameParts.length - 1].charAt(0) + ".";
    },
  },

  props: {
    participant: {
      type: Object,
      required: true
    },
    small: {
      type: Boolean,
      default: false
    },
    client: {
      type: Boolean,
      default: false
    }, edit: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style scoped>
.email-btn {
  position: absolute;
  top: 0;
  right: 0px;
  color: var(--neutral-300);
  border: 1px solid var(--neutral-200);
  padding: 3px 6px;
  font-size: 14px;
  border-radius: 6px;
}

.edit-btn {
  position: absolute;
  top: 0;
  right: 32px;
  color: var(--app_red-500);
  border: 1px solid var(--neutral-200);
  padding: 3px 6px;
  font-size: 14px;
  border-radius: 6px;
  cursor: pointer;
}

</style>