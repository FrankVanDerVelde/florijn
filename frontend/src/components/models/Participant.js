class Participant {

    #id;
    #firstName;
    #lastName;
    #role;
    #avatar;
    #email;

    constructor(id, firstName, lastName, role, avatar, email) {
        this.#id = id;
        this.#firstName = firstName;
        this.#lastName = lastName;
        this.#role = role;
        this.#avatar = avatar;
        this.#email = email;
    }

    get id() {
        return this.#id;
    }

    get firstName() {
        return this.#firstName;
    }

    set firstName(value) {
        this.#firstName = value;
    }

    get lastName() {
        return this.#lastName;
    }

    set lastName(value) {
        this.#lastName = value;
    }

    get role() {
        return this.#role;
    }

    set role(value) {
        this.#role = value;
    }

    get avatar() {
        return this.#avatar;
    }

    set avatar(value) {
        this.#avatar = value;
    }

    get email() {
        return this.#email;
    }

    set email(value) {
        this.#email = value;
    }
}