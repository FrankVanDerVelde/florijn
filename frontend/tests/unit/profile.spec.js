import { mount } from "@vue/test-utils";
import PersonalInfo from '../../src/components/Scenes/Profile/Profile.vue';
import InMemoryEntitiesService from '../mockRepos/InMemoryEntitiesService.js';
import Participant from '../../src/components/models/Participant.js';
let inMemoryUserRepo;

test("has data", () => {
    expect(typeof PersonalInfo.data).toBe("function");
});

test('Profile renders and snapshot is still correct', async () => {
    inMemoryUserRepo = new InMemoryEntitiesService(10000, Participant.createDummy);

    const mockRouter = {
        push: jest.fn()
    }

    const wrapper = mount(PersonalInfo, {
        global: {
            mocks: {
                $router: mockRouter
            },
            provide: {
                "userRepository": inMemoryUserRepo
            },
            stubs: ['FontAwesomeIcon', 'Asset', 'router-link', 'router-view']
        }
    })

    expect(wrapper.isVisible).toBeTruthy();

    expect(mockRouter.push).toHaveBeenCalledTimes(1)

    expect(wrapper.element).toMatchSnapshot()
})