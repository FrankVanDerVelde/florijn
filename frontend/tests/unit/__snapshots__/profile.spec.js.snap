// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders and snapshot is still correct 1`] = `
<div
  class="page-main-mw p-4 md:pt-[5em] flex"
>
  <div
    class="profile-container flex self-center w-full flex-col md:flex-row"
  >
    <div
      class="md:mr-[4em] mb-4 md:mb-0 justify-center md:justify-start"
    >
      
      
    </div>
    <div
      class="grow page-content"
    >
      <router-view-stub />
    </div>
  </div>
</div>
`;

exports[`Profile renders and snapshot is still correct 1`] = `
<div
  class="page-main-mw p-4 md:pt-[5em] flex"
>
  <div
    class="profile-container flex self-center w-full flex-col md:flex-row"
  >
    <div
      class="md:mr-[4em] mb-4 md:mb-0 justify-center md:justify-start"
    >
      
      
    </div>
    <div
      class="grow page-content"
    >
      <router-view-stub />
    </div>
  </div>
</div>
`;
